Include/internal/pycore_critical_section.h

```diff
@@ -0,0 +1,204 @@
+#ifndef Py_INTERNAL_CRITICAL_SECTION_H
+#define Py_INTERNAL_CRITICAL_SECTION_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_lock.h"        // PyMutex
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Implementation of Python critical sections: helpers to replace the global
+// interpreter lock with fine grained locking.
+//
+// A Python critical section is a region of code that can only be executed by
+// a single thread at at time. The regions begin with a call to
+// Py_BEGIN_CRITICAL_SECTION and ends with Py_END_CRITICAL_SECTION. Active
+// critical sections are *implicitly* suspended whenever the thread calls
+// `_PyThreadState_Detach()` (i.e., at tiems when the interpreter would have
+// released the GIL).
+//
+// The most recent critical section is resumed when `_PyThreadState_Attach()`
+// is called. See `_PyCriticalSection_Resume()`.
+//
+// The purpose of implicitly ending critical sections is to avoid deadlocks
+// when locking multiple objects. Any time a thread would have released the
+// GIL, it releases all locks from active critical sections. This includes
+// times when a thread blocks while attempting to acquire a lock.
+//
+// Note that the macros Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION
+// are no-ops in non-free-threaded builds.
+//
+// Example usage:
+//  Py_BEGIN_CRITICAL_SECTION(op);
+//  ...
+//  Py_END_CRITICAL_SECTION;
+//
+// To lock two objects at once:
+//  Py_BEGIN_CRITICAL_SECTION2(op1, op2);
+//  ...
+//  Py_END_CRITICAL_SECTION2;
+
+
+// Tagged pointers to critical sections use the two least significant bits to
+// mark if the pointed-to critical section is inactive and whether it is a
+// _PyCriticalSection2 object.
+#define _Py_CRITICAL_SECTION_INACTIVE       0x1
+#define _Py_CRITICAL_SECTION_TWO_MUTEXES    0x2
+#define _Py_CRITICAL_SECTION_MASK           0x3
+
+#ifdef Py_NOGIL
+#define Py_BEGIN_CRITICAL_SECTION(op) {                             \
+    _PyCriticalSection _cs;                                          \
+    _PyCriticalSection_Begin(&_cs, &_PyObject_CAST(op)->ob_mutex)
+
+#define Py_END_CRITICAL_SECTION                                     \
+    _PyCriticalSection_End(&_cs);                                    \
+}
```

Comment:
What's the behavior if someone forgets to put `Py_END_CRITICAL_SECTION` close it out?  I'm guessing it would be a typical hard-to-decipher-if-you-haven't-experienced-it-before cascade of compiler error due to the missing closing curly brace.  Is that right?  I'm thinking about the possible effect on contributors.

I guess it doesn't matter too much though.  This is essentially the same pattern we follow with `Py_END_ALLOW_THREADS`, which isn't any sort of real burden.

Reply:
Yeah, it's the same as `Py_END_ALLOW_THREADS` (which isn't great). However, the cryptic compile errors are still better than if we didn't have the curly braces and it were a hard-to-debug runtime error.

Reply:
> However, the cryptic compile errors are still better than if we didn't have the curly braces and it were a hard-to-debug runtime error.

I completely agree.